lib/ansible/modules/network/dcnm/dcnm_vrf.py

```diff
@@ -0,0 +1,280 @@
+
+
+def vrf_create_payload(module):
+
+    payload = dict()
+    payload['fabric'] = module.params['fabric']
+    payload['vrfName'] = module.params['vrfName']
+    payload['vrfTemplate'] = module.params['vrfTemplate']
+    payload['vrfExtensionTemplate'] = module.params['vrfExtensionTemplate']
+    payload['vrfTemplateConfig'] = module.params['vrfTemplateConfig']
+    payload['vrfId'] = module.params['vrfId']
+
+    json_data = json.dumps(payload)
+
+    return json_data
+
+
+def vrf_attach_payload(module):
+
+    payload_list = list()
+    payload = dict()
+    payload['vrfName'] = module.params['vrfName']
+    payload['lanAttachList'] = list()
+
+    for serials in module.params['serialNumbersVlans']:
+        for serial, vlan in serials.items():
+            device_payload = dict()
+            device_payload['fabric'] = module.params['fabric']
+            device_payload['vrfName'] = module.params['vrfName']
+            device_payload['serialNumber'] = serial
+            device_payload['vlan'] = vlan
+            device_payload['deployment'] = module.params['deployment']
+            device_payload['extensionValues'] = ""
+            device_payload['instanceValues'] = ""
+            device_payload['freeformConfig'] = ""
+
+        payload['lanAttachList'].append(device_payload)
+
+    payload_list.append(payload)
+    json_data = json.dumps(payload_list)
+
+    return json_data
+
+
+def vrf_deploy_payload(module):
+
+    payload = dict()
+    payload['vrfNames'] = module.params['vrfName']
+
+    json_data = json.dumps(payload)
+
+    return json_data
+
+
+def main():
+    """ main entry point for module execution
+    """
+
+    element_spec = dict(
+        action=dict(required=True, choices=['create', 'attach', 'deploy']),
+        fabric=dict(required=True, type='str'),
+        vrfName=dict(required=True, type='str'),
+        vrfTemplate=dict(default='Default_VRF_Universal', type='str'),
+        vrfExtensionTemplate=dict(default='Default_VRF_Extension_Universal', type='str'),
+        vrfTemplateConfig=dict(default='', type='str'),
+        vrfId=dict(type=int),
+        serialNumbersVlans=dict(type='list'),
+        deployment=dict(default=False, type='bool')
+    )
+
+    required_one_of = [['vrfId', 'vrfName']]
+    mutually_exclusive = [['vrfId'],
+                          ['serialNumbersVlans']]
+
+    module = AnsibleModule(argument_spec=element_spec,
+                           required_one_of=required_one_of,
+                           mutually_exclusive=mutually_exclusive,
+                           supports_check_mode=True)
+
+    result = dict(
+        changed=False,
+        response=dict()
+    )
+
+    action = module.params['action']
+    fabric = module.params['fabric']
+
+    method = 'POST'
+    create_path = '/rest/top-down/fabrics/' + fabric + '/vrfs'
+    attach_path = '/rest/top-down/fabrics/' + fabric + '/vrfs/attachments'
+    deploy_path = '/rest/top-down/fabrics/' + fabric + '/vrfs/deployments'
```

Comment:
nit: consider setting the basic path here and then appending to it in the `if` blocks further down; e.g.

```
    path = '/rest/top-down/fabrics/{}/vrf'.format(fabric)
```

```
    if action == 'create':
           ...
            result['response'] = conn.send_request(method, path, json_data)
          ...
    elif action == 'attach':
           ...
            path += '/attachments
            result['response'] = conn.send_request(method, path, json_data)
          ...

```

Reply:
Can do this, makes it neat...